src/components/join-flow/join-flow.jsx

```diff
@@ -121,21 +121,27 @@ class JoinFlow extends React.Component {
     //   * "birth_month": ["Ensure this value is greater than or equal to 1."]
     handleRegistrationResponse (err, body, res) {
         this.setState({
-            numAttempts: this.state.numAttempts + 1,
-            waiting: false
+            numAttempts: this.state.numAttempts + 1
         }, () => {
             const success = this.registrationIsSuccessful(err, body, res);
             if (success) {
-                this.props.refreshSession();
-                this.setState({step: this.state.step + 1});
+                this.props.refreshSessionWithRetry().then(() => {
+                    this.setState({
+                        step: this.state.step + 1,
+                        waiting: false
+                    });
+                });
                 return;
             }
             // now we know something went wrong -- either an actual error (client-side
             // or server-side), or just a problem with the registration content.
 
             // if an actual error, prompt user to try again.
             if (err || res.statusCode !== 200) {
-                this.setState({registrationError: {errorAllowsTryAgain: true}});
+                this.setState({
+                    registrationError: {errorAllowsTryAgain: true},
+                    waiting: false
```

Comment:
Since I took out the `setState({...waiting:false})` above, I have to be sure to set it in all these cases, or the user will be stuck with a spinner forever...